### PR TITLE
PROTOTECH 49: Revert nicely if no `decimals()` method exists QT or CT

### PR DESCRIPTION
**Pull request by ith-harvey**
# Description of change
## High level
* Created a `hasDecimalsMethod()` method which checks if a contract has a `decimals()` method
* Created a `TokenInvalidNoDecimals` revert if a contract doesn't have the `decimals()` method

# Description of bug or vulnerability and solution
* Original Prototech issue: https://github.com/Fixed-Point-Solutions/prototech-ajna-audit/issues/49
* Although we point out that the `decimals()` method isn't required in the readme it wasn't reverting cleanly. 

# Contract size
## Pre Change
```
============ Deployment Bytecode Sizes ============
  ERC721Pool               -  24,386B  (99.22%)
  ERC20Pool                -  23,614B  (96.08%)
  PositionManager          -  19,302B  (78.54%)
  PositionNFTSVG           -  15,257B  (62.08%)
  TakerActions             -  15,104B  (61.46%)
  PoolInfoUtils            -  13,622B  (55.43%)
  BorrowerActions          -  12,357B  (50.28%)
  LenderActions            -  10,581B  (43.05%)
  KickerActions            -  10,435B  (42.46%)
  SettlerActions           -   9,270B  (37.72%)
  PoolCommons              -   9,082B  (36.95%)
  RewardsManager           -   8,522B  (34.67%)
  ERC721                   -   4,276B  (17.40%)
  ERC721PoolFactory        -   3,361B  (13.68%)
  LPActions                -   2,952B  (12.01%)
  ERC20PoolFactory         -   2,602B  (10.59%)
  ERC20                    -   2,114B  (8.60%)
  Address                  -      86B  (0.35%)
  Base64                   -      86B  (0.35%)
  Buckets                  -      86B  (0.35%)
  ClonesWithImmutableArgs  -      86B  (0.35%)
  Deposits                 -      86B  (0.35%)
  ECDSA                    -      86B  (0.35%)
  EnumerableSet            -      86B  (0.35%)
  Loans                    -      86B  (0.35%)
  Math                     -      86B  (0.35%)
  Maths                    -      86B  (0.35%)
  PRBMath                  -      86B  (0.35%)
  PRBMathSD59x18           -      86B  (0.35%)
  PRBMathUD60x18           -      86B  (0.35%)
  SafeCast                 -      86B  (0.35%)
  SafeERC20                -      86B  (0.35%)
  SignatureChecker         -      86B  (0.35%)
  Strings                  -      86B  (0.35%)
  Clone                    -      63B  (0.26%)
```
## Post Change
```
============ Deployment Bytecode Sizes ============
  ERC721Pool               -  24,386B  (99.22%)
  ERC20Pool                -  23,614B  (96.08%)
  PositionManager          -  19,171B  (78.00%)
  PositionNFTSVG           -  15,257B  (62.08%)
  TakerActions             -  15,104B  (61.46%)
  PoolInfoUtils            -  13,622B  (55.43%)
  BorrowerActions          -  12,357B  (50.28%)
  LenderActions            -  10,581B  (43.05%)
  KickerActions            -  10,435B  (42.46%)
  SettlerActions           -   9,270B  (37.72%)
  PoolCommons              -   9,082B  (36.95%)
  RewardsManager           -   8,522B  (34.67%)
  ERC721                   -   4,276B  (17.40%)
  ERC721PoolFactory        -   3,615B  (14.71%)
  LPActions                -   2,952B  (12.01%)
  ERC20PoolFactory         -   2,904B  (11.82%)
  ERC20                    -   2,114B  (8.60%)
  Address                  -      86B  (0.35%)
  Base64                   -      86B  (0.35%)
  Buckets                  -      86B  (0.35%)
  ClonesWithImmutableArgs  -      86B  (0.35%)
  Deposits                 -      86B  (0.35%)
  EnumerableSet            -      86B  (0.35%)
  Loans                    -      86B  (0.35%)
  Math                     -      86B  (0.35%)
  Maths                    -      86B  (0.35%)
  PRBMath                  -      86B  (0.35%)
  PRBMathSD59x18           -      86B  (0.35%)
  PRBMathUD60x18           -      86B  (0.35%)
  SafeCast                 -      86B  (0.35%)
  SafeERC20                -      86B  (0.35%)
  Strings                  -      86B  (0.35%)
  Clone                    -      63B  (0.26%)
```

# Gas usage
## Pre Change
<PASTE_OUTPUT_HERE>
## Post Change
<PASTE_OUTPUT_HERE>

### CODE REVIEW

File: tests/forge/unit/ERC721Pool/ERC721PoolFactory.t.sol [hunk 1]
```diff
@@ -110,19 +112,32 @@ contract ERC721PoolFactoryTest is ERC721HelperContract {
...
+            interestRate: 0.05 * 10**18
+        });
+    }
+
     function testDeployERC721CollectionPoolWithInvalidRate() external {
         // should revert if trying to deploy with interest rate lower than accepted
         _assertDeployWithInvalidRateRevert({
             poolFactory:  address(_factory),
-            collateral:   address(new NFTCollateralToken()),
+            collateral:   address(_collateral),
```
Comment (ith-harvey) [inline tests/forge/unit/ERC721Pool/ERC721PoolFactory.t.sol:132]: `_collateral` was already equal to `NFTCollateralToken()` above, this is cleanup

File: src/base/PoolDeployer.sol [hunk 2]
```diff
@@ -67,4 +67,11 @@ abstract contract PoolDeployer {
         return deployedPoolsList.length;
     }
 
+    function hasDecimalsMethod(address contract_) internal view returns (bool methodExists_) {
+        // 0x313ce567 = bytes4(keccak256("decimals()""))
+        (methodExists_,) = contract_.staticcall(abi.encodePacked(bytes4(0x313ce567)));
```
Comment (EdNoepel) [inline src/base/PoolDeployer.sol:72]: Cool technique; I was unaware you could do that.

**Review by EdNoepel — APPROVED** [review]
Nice; LGTM

File: src/base/PoolDeployer.sol [hunk 3]
```diff
@@ -67,4 +67,11 @@ abstract contract PoolDeployer {
...
+    }
+
+
```
Comment (MikeHathaway) [inline src/base/PoolDeployer.sol]: nit: extra space

File: tests/forge/utils/ContractERC20NoDecimals.sol [hunk 4]
```diff
@@ -0,0 +1,399 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts (last updated v4.8.0) (token/ERC20/ERC20.sol)
+
+pragma solidity ^0.8.0;
+
+// import "./IERC20.sol";
+// import "./extensions/IERC20Metadata.sol";
+// import "../../utils/Context.sol";
```
Comment (MikeHathaway) [inline tests/forge/utils/ContractERC20NoDecimals.sol]: can remove unused imports

**Review by MikeHathaway — APPROVED** [review]
LGTM. Just a few minor comments

File: tests/forge/utils/ContractERC20NoDecimals.sol (hunk 4 above)
Comment (ith-harvey) [reply tests/forge/utils/ContractERC20NoDecimals.sol]: resolved -> https://github.com/ajna-finance/contracts/pull/864/commits/fc03fb5e2ebd23d4b9b5be1f9454ced92f7b5113

File: src/base/PoolDeployer.sol (hunk 3 above)
Comment (ith-harvey) [reply src/base/PoolDeployer.sol]: resolved -> https://github.com/ajna-finance/contracts/pull/864/commits/fc03fb5e2ebd23d4b9b5be1f9454ced92f7b5113

File: src/base/PoolDeployer.sol [hunk 5]
```diff
@@ -67,4 +67,8 @@ abstract contract PoolDeployer {
         return deployedPoolsList.length;
     }
 
+    function hasDecimalsMethod(address contract_) internal view returns (bool methodExists_) {
+        // 0x313ce567 = bytes4(keccak256("decimals()""))
+        (methodExists_,) = contract_.staticcall(abi.encodePacked(bytes4(0x313ce567)));
```
Comment (grandizzy) [inline src/base/PoolDeployer.sol:72]: Wouldn't be better to have this entire abi.encoded as a constant instead computing each time?

Comment (prat-gpt) [reply src/base/PoolDeployer.sol:72]: yes right, we can encode is as a constant.

**Pull request merged**